gbpservice/nfp/lifecycle_manager/modules/service_lifecycle_manager.py

```diff
@@ -0,0 +1,661 @@
+            request_data['network_function_id'],
+            updated_network_function)
+        # Trigger RPC to notify the Create_Service caller with status
+
+    # TODO: When Device LCM deletes Device DB, the Foreign key NSI will be nulled
+    def handle_user_config_deleted(self, event):
+        request_data = event.data
+        network_function = self.db_handler.get_network_function(
+            self.db_session,
+            request_data['network_function_id'])
+        for nfi_id in network_function['network_function_instances']:
+            self._create_event('DELETE_NETWORK_FUNCTION_INSTANCE',
+                               event_data=nfi_id)
+
+    # Change to Delete_failed or continue with instance and device
+    # delete if config delete fails? or status CONFIG_DELETE_FAILED ??
+    def handle_user_config_delete_failed(self, event):
+        request_data = event.data
+        updated_network_function = {
+            'status': 'ERROR',
+        }
+        self.db_handler.update_network_function(
+            self.db_session,
+            request_data['network_function_id'],
+            updated_network_function)
+        # Trigger RPC to notify the Create_Service caller with status ??
+
+    # When Device LCM deletes Device DB, the Foreign key NSI will be nulled
+    # So we have to pass the NSI ID in delete event to device LCM and process
+    # the result based on that
+    def handle_device_deleted(self, event):
+        request_data = event.data
+        nfi_id = request_data['network_function_instance_id']
+        nfi = self.db_handler.get_network_function_instance(
+            self.db_session, nfi_id)
+        self.db_handler.delete_network_function_instance(
+            self.db_session, nfi_id)
+        network_function = self.db_handler.get_network_function(
+            self.db_session, nfi['network_function_id'])
+        if not network_function['network_function_instances']:
+            self.db_handler.delete_network_function(
+                self.db_session, nfi['network_function_id'])
+            # Inform delete service caller with delete completed RPC
+
+    # But how do you get hold of the first event ??
+    def _request_completed(self, event):
+        self._controller.event_done(event)
+
+    def get_network_function(self, context, network_function_id):
+        try:
+            network_function = self.db_handler.get_network_function(
+                self.db_session, network_function_id)
+            LOG.info(_("In get_network_function, returning: %(network_function)s"), {'network_function': network_function})
+            return network_function
+        except Exception:
+            LOG.exception(_("Error in get_network_function"))
+            return None
+
+    def get_network_functions(self, context, filters):
+        return self.db_handler.get_network_functions(
+            self.db_session, filters)
+
+    def handle_policy_target_added(self, context, network_function_id,
```

Comment:
Need to discuss. You might have an alternate mechanism to abstract out GBP and neutron.
1. Similar to DLCM, we can have SLCM drivers slcm_gbp_service.py, slcm_neutron_service.py (alternate names?)
2. Method here will call the driver method for driver specific impl. For eg, this method in neutron will be a pass.
3. post next event etc will be done device_lc_manager.py and not in the drivers.
4. All the gbp code can move into the driver.
5. gbp, neutron drivers to make any api calls will use lifecycle_management/networking/api.py class
6. get_service_details impl in this file will also move into the driver.